README.md
# linear-regression

Clojure linear regression implementation

## Installation

```
git clone https://github.com/ftriquet/ft_linear_regression.git
```
```
lein uberjar
```

## Usage

    $ java -jar target/uberjar/linear-regression-0.1.0-standalone.jar [args]

## Options

--predict: Prompt for values and outputs the predictions based on previously generated model

--file FILE: specifies the dataset file. The file must be csv formated

--train: Start training on a dataset specified by the FILE option

--graph: Display the calculated graph if the regression has one parameter

--animation: Display an animation of the graph fitting the dataset if the regression has one parameter


## Examples

    $ java -jar target/uberjar/linear-regression-0.1.0-standalone.jar --train --animation --file data.csv


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-08-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-08-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/linear-regression/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/linear-regression/compare/0.1.0...0.1.1


test/linear_regression/core_test.clj
(ns linear-regression.core-test
  (:require [clojure.test :refer :all]
            [linear-regression.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 1 1))))


test/linear_regression/matrix_test.clj
(ns linear-regression.matrix-test
  (:require [clojure.test :refer :all]
            [linear-regression.matrix :refer :all]))

(deftest l-vec?-test
  (is (l-vec? [1 2 3 4]) "Line vector")
  (is (= false (l-vec? 1)) "Number")
  (is (= false (l-vec? [[1] [2]])) "Column vector")
  (is (= false (l-vec? [[1 2 3] [4 5 6] [7 8 9]])) "Matrix"))

(deftest c-vec?-test

  (is (= false (c-vec? [1 2 3 4])) "Line vector")
  (is (= false (c-vec? 1)) "Number")
  (is (= true  (c-vec? [[1] [2]])) "Column vector")
  (is (= false (c-vec? [[1 2 3] [4 5 6] [7 8 9]])) "Matrix"))


(deftest matrix?-test
  (is (= true (matrix? [[1 4] [2 5] [3 6]])) "Matrix")
  (is (= false (matrix? 8)) "Number")
  (is (= false (matrix? [1 2 3 4])) "Vector")
  (is (= true (matrix? [[1] [2] [3]])) "Column vector"))

(deftest dims-test
  (is (= '(1 2) (dims [4 5])))
  (is (= '(3 4) (dims [[1 2 3 4] [1 2 3 4] [1 2 3 4]]))))


(deftest mult-test
  (is (= 4 (mult 2 2)) "Numbers")

  (is (= [[6 10]
          [8 12]]
         (mult 2 [[3 5] [4 6]])) "Number and matrix")

  (is (= [[6 10]
          [8 12]]
         (mult [[3 5] [4 6]] 2)) "Matrix and number")

  (is (= [2 4 6 8 10] (mult 2 [1 2 3 4 5])) "Number and line vector")

  (is (= [[2] [4] [6] [8] [10]] (mult 2 [[1] [2] [3] [4] [5]])) "Number and column vector")

  (is (= [15 70] (mult [1 2 3 4 5] [[1 2]
                                    [1 3]
                                    [1 4]
                                    [1 5]
                                    [1 6]])))

  (is (= [43 43] (mult [1 2 3 4] [[7 7] [3 3] [2 2] [6 6]])))
  (is (= [[30 36 42]
          [66 81 96]
          [102 126 150]]
         (mult [[1 2 3]
                [4 5 6]
                [7 8 9]]
               [[1 2 3]
                [4 5 6]
                [7 8 9]])) "Two square matrices")

  (is (= (mult [1 2] [[1 100] [1 100]]) [3 300]) "Line by matrix")
  (is (= 30 (mult [1 2 3 4] [[1] [2] [3] [4]])) "Line by column")
  (is (= (mult [[1] [2]] [1 100])
         [[1 100]
          [2 200]]) "Column by line")

  (is (= [[14] [32]] (mult [[1 2 3] [4 5 6]] [[1] [2] [3]])) "Matrix by column vector"))

(deftest transpose-test
  (is (= (transpose [[1 2 3 4]]) [[1] [2] [3] [4]]) "Line vector")
  (is (= (transpose [[1 2 3]
                     [4 5 6]]) [[1 4]
                                [2 5]
                                [3 6]]) "Matrix")
  (is (= (transpose [[1] [2] [3] [4]]) [1 2 3 4]) "Column vector"))


(deftest preds-test
  (is (= [38.0 16.0 15.0] (preds [[1.0 3.0 5.0 8.0]
                                  [1.0 4.0 1.0 3.0]
                                  [1.0 7.0 2.0 1.0]]
                                 [1.0 1.0 2.0 3.0]))))

(deftest costs-test
  (let [inputs [[1.0 3.0 5.0 8.0]
                [1.0 4.0 1.0 3.0]
                [1.0 7.0 2.0 1.0]]
        weights [1.0 1.0 2.0 3.0]
        labels [40.0 12.0 13.0]
        expected [4.0 16.0 4.0]]

    (is (= expected (costs inputs weights labels #(* % %))))))


test/linear_regression/learning_test.clj
(ns linear-regression.learning-test
  (:require [clojure.test :refer :all]
            [linear-regression.learning :refer :all]))

(deftest scale-test
  (is (= (scale [[1.0 13.0 3.0 2.0]
                 [13.0 4.0 8.0 6.0]
                 [25.0 28.0 6.0 6.0]])
         [[0.0 0.375 0.0 0.0]
          [0.5 0.0 1.0 1.0]
          [1.0 1.0 0.6 1.0]]))

  (is (= (scale [[2.0 26.0 6.0 4.0]
                 [26.0 8.0 16.0 12.0]
                 [50.0 56.0 12.0 12.0]])
         [[0.0 0.375 0.0 0.0]
          [0.5 0.0 1.0 1.0]
          [1.0 1.0 0.6 1.0]]))

  )


src/linear_regression/matrix.clj
(ns linear-regression.matrix)

(defn l-vec? [m]
  (and (coll? m) (not (coll? (first m)))))

(defn c-vec? [m]
  (if (not (number? m))
    (let [f (first m)]
      (and (coll? f) (= 1 (count f))))
    false))

(defn matrix? [m]
  (and (not (number? m)) (not (l-vec? m))))

(defn mprint [m]
  (cond
    (number? m) (println m)
    (l-vec? m) (println m)
    :else (doseq [l m] (println l))))

(defn transpose [m]
  (cond
    (number? m) m
    (c-vec? m) (flatten m)
    (l-vec? m) (map list m)
    :else (apply map list m)))

(defn dims [m]
  (cond
    (number? m) 0
    (l-vec? m) (list 1 (count m))
    :else (list (count m) (count (first m)))))

(defn lines [m]
  (first (dims m)))

(defn cols [m]
  (second (dims m)))

(defn to-line [v]
  (if (l-vec? v)
    v
    (transpose v)))

(defn to-col [v]
  (if (c-vec? v)
    v
    (transpose v)))


(defn mult [m1 m2]
  (cond
    (number? m1) (if (number? m2)
                   (* m1 m2)
                   (map #(mult m1 %) m2))
    (number? m2) (mult m2 m1)
    (not= (lines m2)  (cols m1)) (throw (IllegalArgumentException.
                                          (str "Invalid matrix multiplication: " (lines m1) "x" (cols m1) "*" (lines m2) "x" (cols m2))))
    (l-vec? m1) (if (c-vec? m2)
                  (reduce + (map * m1 (transpose m2)))
                  (for [col (transpose m2)]
                    (reduce + (map * m1 col))))
    (c-vec? m1) (if (l-vec? m2)
                  (for [line m1]
                    (mult (first line) m2))
                  (let [tr (transpose m1)]
                    (for [i m2]
                      (transpose (map #(* i %) tr)))))
    (c-vec? m2) (for [line m1]
                  (list (mult line m2)))
    :else (partition (count m1)
                     (for [line m1 col (transpose m2)]
                       (reduce + (map * line col))))))
(defn sub [m1 m2]
  (cond
    (and (l-vec? m1) (l-vec? m2)) (map - m1 m2)
    (= (dims m1) (dims m2)) (map sub m1 m2)
    :else (throw (IllegalArgumentException.  "Invalid matrix substraction"))))

(defn add [m1 m2]
  (cond
    (and (l-vec? m1) (l-vec? m2)) (map + m1 m2)
    (= (dims m1) (dims m2)) (map add m1 m2)
    :else (throw (IllegalArgumentException.  "Invalid matrix addition"))))

(defn add-bias-feature [inputs]
 (if (l-vec? inputs)
   (cons 1.0 inputs)
   (map add-bias-feature inputs)))

(defn preds [inputs weights]
  (to-line
    (if (l-vec? inputs)
      (reduce + (map * inputs (to-line weights)))
      (mult inputs (to-col weights)))))

(defn abs [x]
  (if (< 0 x) (- x) x))

(defn costs [inputs weights labels f]
  (let [predictions (preds inputs weights)]
    (map #(f  (- %1 %2)) (to-line predictions) (to-line labels))))

(defn cost [inputs weights labels f]
  (/ (float (reduce + (costs inputs weights labels f))) (* 2 (lines inputs))))



src/linear_regression/learning.clj
(ns linear-regression.learning
  (:require [linear-regression.matrix :refer :all]
            [quil.core :as q]))

(defn sq [x]
  (* x x))

(defn zeros [n]
  (take n (repeat 0.0)))

(defn f-range [v]
  (-  (apply max v) (apply min v)))

(defn f-scale [inputs f]
  (if (not (c-vec? inputs))
    (transpose (for [col (transpose inputs)]
                 (let [rng (f-range col)
                       a (f col)]
                   (map #(/ (- % a) rng) col))))
    (let [l (transpose inputs)
          rng (f-range l)
          a (f l)]
      (transpose (map #(/ (- % a) rng) l)))))

(defn scale [inputs]
  (f-scale inputs (partial apply min)))


(defn draw-points [{X :X labels :labels}]
  (q/background 255)
  (doseq [[x [y]] (partition 2 (interleave (map second X) labels))]
    (q/rect (* 800.0 x)
            (- 800 (* 800.0 (/ (double y) (apply max (map first labels)))))
            4 4)))

(defn compute-costs [X theta labels]
  (sub (mult X (to-col theta)) labels))


(defn get-ext [inputs]
  (let [tr (transpose inputs)]
    (if (l-vec? tr)
      (list (list (apply min tr)) (list (apply max tr)))
      (list (map #(apply min %) tr) (map #(apply max %) tr)))))

(defn init-draw [{X :X theta :theta labels :labels minx :min maxx :max :as state}]
  (fn []
    (q/frame-rate 1)
    (q/background 240)
    {:X X
     :min minx
     :max maxx
     :theta theta
     :labels labels}))

(defn upd [state] state)

(defn draw-result [{X :X theta :theta labels :labels minx :min maxx :max :as state}]
  (q/background 255)
  (doseq [[x [y]] (partition 2 (interleave (map second X) labels))]
    (q/rect (* 800.0 x)
            (- 800 (* 800.0 (/ (double y) (apply max (map first labels)))))
            4 4))
  (let [x1 0.0
        y1 (+ (first theta) (* x1 (second theta)))
        x2 1.0
        y2 (+ (first theta) (* x2 (second theta)))]
    (q/line
      (* 800.0 x1)
      (- 800 (* 800.0 (/ (double y1) (apply max (map first labels)))))
      (* 800.0 x2)
      (- 800 (* 800.0 (/ (double y2) (apply max (map first labels))))))))


(defn train-model [inputs labels lrn-rate]
  (let [X (add-bias-feature (scale inputs))
        weights (zeros (inc (cols inputs)))
        [minsX maxsX] (get-ext inputs)]
    (println "Start computing, actual thetas: " weights)
    (loop [theta weights
           error (cost X theta labels sq)]
      (let [d-theta (mult (/ lrn-rate (double (lines X)))
                          (mult (to-line (compute-costs X theta labels)) X))
            new-theta (sub theta d-theta)
            err (cost X new-theta labels sq)]
        (if (< (- error err) 0.000001)
          (do
            (println "Done! Thetas: " new-theta)
            {:csv-data [["mins" "maxs" "theta"]
                        [(pr-str minsX) (pr-str maxsX) (pr-str new-theta)]]
             :X X
             :theta new-theta
             :labels labels
             :min minsX
             :max maxsX})
          (recur new-theta err))))))


(defn init-st [{X :X labels :labels :as state}]
  (fn []
    (let [[minsX maxsX] (get-ext X)]
      (q/frame-rate 60)
      (q/background 240)
      {:X (add-bias-feature (scale X))
       :min minsX
       :max maxsX
       :theta (zeros 2)
       :labels labels})))


(defn update-state [{X :X theta :theta labels :labels :as state}]
  (let [d-theta (mult (/ 0.1 (double (lines X)))
                      (mult (to-line (compute-costs X theta labels)) X))
        new-theta (sub theta d-theta)]
    (assoc state :theta new-theta)))


src/linear_regression/core.clj
(ns linear-regression.core
  (:gen-class)
  (:require [linear-regression.matrix :refer :all]
            [linear-regression.learning :refer :all]
            [quil.core :as q]
            [quil.middleware :as m]
            [clojure.tools.cli :as cli]
            [clojure.java.io :as io]
            [clojure.string :as string]
            [clojure-csv.core :as csv]))

(def cli-options [["-t" "--train"]
                  ["-p" "--predict"]
                  ["-g" "--graph"]
                  ["-a" "--animation"]
                  ["-f" "--file FILE" "data file" :default "resources/data.csv"]])

(defn- check-data [data]
  (let [s (count (first data))]
    (when (every? #(= (count %) s) data) true)))

(defn- format-data [data]
  (let [to-f (map #(map (comp float read-string) %) data)]
    (list (map (partial drop-last 1) to-f)
          (map (comp list last) to-f))))


(defn get-data-from-csv []
  (let [csv-data (second (csv/parse-csv (slurp "resources/thetas.csv")))
        [mins maxs theta] (map read-string csv-data)]
    (println "Enter value(s)")
    (try
      (let [input (map read-string (string/split (read-line) #" "))]
        (if (not= (count input) (dec (count theta)))
          (println "Invalid arguments number")
          (println (mult (->> mins
                              (map - input)
                              (map #(/ %3 (- %1 %2)) maxs mins)
                              (cons 1.0))
                         (to-col theta)))))
      (catch Exception e))))

(defn predict []
  (let [predict-data (if (.exists (io/as-file "resources/thetas.csv"))
                       (get-data-from-csv)
                       {:min 0 :max 1 :theta [0 0]})]))

(defn train-theta [{animation :animation file :file graph :graph}]
  (let [file-content (try (slurp file) (catch Exception e nil))]
    (if (nil? file-content)
      (println "File not found")
      (let [csv-data (csv/parse-csv file-content)
            columns (first csv-data)
            data (rest csv-data)]
        (if (check-data data)
          (let [[inputs labels] (format-data data)
                results (train-model inputs labels 0.1)]
            (->> (:csv-data results)
                 (csv/write-csv)
                 (spit "resources/thetas.csv"))
            (when (= 2 (cols (:X results)))
              (cond
                animation (q/defsketch linear-regression
                            :title "Linear Regression"
                            :size [800 800]
                            :setup (init-st {:X inputs :labels labels})
                            :draw draw-result
                            :update update-state
                            :middleware [m/fun-mode])
                graph     (q/defsketch linear-regression
                            :title "Linear Regression"
                            :size [800 800]
                            :setup (init-draw results)
                            :draw draw-result
                            :update upd
                            :middleware [m/fun-mode]))))
          (println "Invalid csv file"))))))


(defn -main [& args]
  (let [opts (cli/parse-opts args cli-options)
        options (:options opts)]
    (if (:errors opts) (doseq [x (list "Usage: " (:summary opts))] (println x))
      (cond
        (:train options) (train-theta options)
        (:predict options) (predict)
        :else (println "Invalid arguments")))))
